### PR TITLE
server: return usage description with resource names and UUIDs

### PR DESCRIPTION
**Pull request by rhtyd**
The usage record descriptions have CloudStack's internal integer IDs
which makes it difficult for users to read their usages. This PRs
introduces a new API boolean flag `oldformat` which when set to true
would return the older description format, otherwise by default
listUsageRecords will process and return description with names and
UUIDs of resources.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java [hunk 1]
```diff
@@ -81,6 +82,10 @@
     @Parameter(name = ApiConstants.INCLUDE_TAGS, type = CommandType.BOOLEAN, description = "Flag to enable display of Tags for a resource")
     private Boolean includeTags;
 
+    @Parameter(name = ApiConstants.OLD_FORMAT, type = CommandType.BOOLEAN, description = "Flag to enable description rendered in old format with no processing")
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: should this have a default (of true)?

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java [hunk 2]
```diff
@@ -145,6 +150,9 @@ public void setUsageId(String usageId) {
         this.usageId = usageId;
     }
 
+    public boolean getOldFormat() {
+        return oldFormat != null && oldFormat;
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: this is a way to make the default be false. It should be in the description of the Parameter-annotation.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java (hunk 1 above)
Comment (andrijapanicsb) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: 1. old behaviour should be the default, unless the documentation is (somewhere...) updated.
2. word "processing" - does this mean that there will be additional CPU usage and DB hammering on top of existing (I know some people's usage_jobs taking many hours to complete - would like to understand performance consequences)

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java (hunk 2 above)
Comment (rhtyd) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: Will fix description.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java (hunk 1 above)
Comment (rhtyd) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: I'll update the docs. Will fix description.

**Comment by rhtyd** [comment]
@DaanHoogland @andrijapanic I've sent the docs PR https://github.com/apache/cloudstack-documentation/pulls
And updated the API doc description

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java (hunk 2 above)
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: I agree that this default is sensible and I don't think the backwards incompatibility of not using internal ids anymore is going to be very hurtful as the processing of such ids is more intensive than processing uuids using the api. maybe a point of discussion for some, so let's be extra cautious.

**Comment by andrijapanicsb** [comment]
LGTM

Possible consideration is the performance penalty due to API having to query the main "cloud" DB.
If no crazy JOINs, than it should be fine.

@rhtyd @DaanHoogland ^^^

**Comment by rhtyd** [comment]
@andrijapanic thanks, sounds good to me. This is a given that using UUIDs in the description will incur some processing. In large environments this will make the list API slightly slower than before, however all the UUID translation logic makes individual listing (select * query) so no crazy joins but multiple select queries may be executed depending on the type of the record. In few cases I've optimised not querying for a VM/volume/template resource which is already queried in the beginning on the block.

**Comment by andrijapanicsb** [comment]
Sounds good @rhtyd

LGTM.

**Pull request merged**